redis/src/commands/cluster_scan.rs

```diff
@@ -0,0 +1,709 @@
+    pub(crate) scan_state_cursor: ScanStateCursor,
+    match_pattern: Option<String>,
+    count: Option<usize>,
+    object_type: Option<ObjectType>,
+}
+
+#[derive(Debug, Clone)]
+/// Represents the type of an object in Redis.
+pub enum ObjectType {
+    /// Represents a string object in Redis.
+    STRING,
+    /// Represents a list object in Redis.
+    LIST,
+    /// Represents a set object in Redis.
+    SET,
+    /// Represents a sorted set object in Redis.
+    ZSET,
+    /// Represents a hash object in Redis.
+    HASH,
+}
+
+impl std::fmt::Display for ObjectType {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            ObjectType::STRING => write!(f, "string"),
+            ObjectType::LIST => write!(f, "list"),
+            ObjectType::SET => write!(f, "set"),
+            ObjectType::ZSET => write!(f, "zset"),
+            ObjectType::HASH => write!(f, "hash"),
+        }
+    }
+}
+
+impl ClusterScanArgs {
+    pub(crate) fn new(
+        scan_state_cursor: ScanStateCursor,
+        match_pattern: Option<String>,
+        count: Option<usize>,
+        object_type: Option<ObjectType>,
+    ) -> Self {
+        Self {
+            scan_state_cursor,
+            match_pattern,
+            count,
+            object_type,
+        }
+    }
+}
+
+#[derive(Debug, Clone, Default)]
+/// A wrapper struct for managing the state of a scan operation in a cluster.
+pub struct ScanStateCursor {
+    scan_state_ref: Arc<Option<ScanState>>,
+}
+
+impl ScanStateCursor {
+    /// Creates a new instance of `ScanStateCursor` from a given `ScanState`.
+    fn from_scan_state(scan_state: &ScanState) -> Self {
+        Self {
+            scan_state_ref: Arc::new(Some(scan_state.clone())),
+        }
+    }
+
+    /// Creates a new instance of `ScanStateCursor`.
+    ///
+    /// This method initializes the `ScanStateCursor` with a reference to a `ScanState` that is initially set to `None`.
+    /// An empty ScanState is equivilant to a 0 cursor.
+    pub fn new() -> Self {
+        Self {
+            scan_state_ref: Arc::new(None),
+        }
+    }
+
+    /// Returns `true` if the scan state is empty, indicating that the scan operation has finished or has not started.
+    pub fn is_none(&self) -> bool {
+        self.scan_state_ref.is_none()
+    }
+
+    /// Returns a clone of the scan state, if it exist.
+    pub(crate) fn get_state_from_wrraper(&self) -> Option<ScanState> {
+        if self.is_none() {
+            None
+        } else {
+            self.scan_state_ref.as_ref().clone()
+        }
+    }
+}
+
+/// This trait defines the methods for interacting with a Redis cluster during scanning.
+#[async_trait]
+pub(crate) trait ClusterInScan {
+    /// Retrieves the address address associated with a given slot in the cluster.
+    async fn get_address_by_slot(&self, slot: u16) -> RedisResult<String>;
+
+    /// Retrieves the epoch of a given address in the cluster.
```

Comment:
This comment does not really explain what the EPOCH means in this context, I could write this comment just by reading the function name :)